File: CoreScala/simple-scala/src/main/scala/org/gb/sample/scala/hackerrank/btree/BinaryTree.scala
package org.gb.sample.scala.hackerrank.btree

import scala.math.max

class Node[T](val value: T, val leftChild: Node[T], val rightChild: Node[T]) {
    
    override def toString(): String = {
        String.valueOf(value)
    }
}

case class BinaryTree[T](private val root: Node[T]) {

    private def getHeight(node: Node[T]): Int = {
        if (node == null) {
            return -1
        } else {
            val subtree = BinaryTree(node)
            return subtree.getHeight()
        }
    }

    def getHeight(): Int = {
        val leftSubtreeHeight = getHeight(root.leftChild)
        val rightSubtreeHeight = getHeight(root.rightChild)

        val maxHeightOfChildren = max(leftSubtreeHeight, rightSubtreeHeight)
        return maxHeightOfChildren + 1
    }

    def getRoot(): Node[T] = {
        return root
    }

    def getLeftSubtree(): BinaryTree[T] = {
        if (root.leftChild == null) {
            null
        } else {
            BinaryTree(root.leftChild)
        }
    }

    def getRightSubtree(): BinaryTree[T] = {
        if (root.rightChild == null) {
            null
        } else {
            BinaryTree(root.rightChild)
        }
    }

    override def toString(): String = {
        "root: ".concat(String.valueOf(root.value))
    }

}

File: CoreJava/algorithms/src/main/java/org/gb/sample/algo/hackerrank/btree/BinaryTree.java
package org.gb.sample.algo.hackerrank.btree;

public class BinaryTree<T> {

	private Node<T> root;

	public BinaryTree(Node<T> root) {
		this.root = root;
	}

	int getHeight() {
		int leftSubtreeHeight = getHeight(root.getLeftChild());
		int rightSubtreeHeight = getHeight(root.getRightChild());

		int maxHeightOfChildren = Integer.max(leftSubtreeHeight, rightSubtreeHeight);
		return maxHeightOfChildren + 1;
	}

	private int getHeight(Node<T> node) {
		if (node == null) {
			return -1;
		} else {
			BinaryTree<T> subtree = new BinaryTree<>(node);
			return subtree.getHeight();
		}
	}

	public BinaryTree<T> getLeftSubtree() {
		if (root.getLeftChild() == null) {
			return null;
		} else {
			return new BinaryTree<>(root.getLeftChild());
		}
	}

	public BinaryTree<T> getRightSubtree() {
		if (root.getRightChild() == null) {
			return null;
		} else {
			return new BinaryTree<T>(root.getRightChild());
		}
	}

	@Override
	public String toString() {
		return "root: ".concat(String.valueOf(root.getValue()));
	}

	public Node<T> getRoot() {
		return root;
	}
}
